airbyte-integrations/connectors/source-instagram/source_instagram/client/api.py

```diff
@@ -0,0 +1,268 @@
+        This is necessary because the functions that call this endpoint return
+        a generator, whose calls need decorated with a backoff.
+        """
+        return self._api.account.get_media(params=params, fields=fields)
+
+    def _get_single_record(self, media_id, fields) -> IGMedia:
+        return IGMedia(media_id).api_get(fields=fields)
+
+
+class StoriesAPI(StreamAPI):
+    def list(self, fields: Sequence[str] = None) -> Iterator[dict]:
+        media = self._get_stories({"limit": self.result_return_limit}, fields)
+        for record in media:
+            record_data = record.export_all_data()
+            yield record_data
+
+    def _get_stories(self, params, fields: list) -> Iterator[Any]:
+        """
+        This is necessary because the functions that call this endpoint return
+        a generator, whose calls need decorated with a backoff.
+        """
+        return self._api.account.get_stories(params=params, fields=fields)
+
+
+class MediaInsightsAPI(MediaAPI):
+    MEDIA_METRICS = ["engagement", "impressions", "reach", "saved"]
+    CAROUSEL_ALBUM_METRICS = ["carousel_album_engagement", "carousel_album_impressions", "carousel_album_reach", "carousel_album_saved"]
+
+    def list(self, fields: Sequence[str] = None) -> Iterator[dict]:
+        media = self._get_media({"limit": self.result_return_limit}, ["media_type"])
+        for ig_media in media:
+            yield {
+                **{"id": ig_media.get("id")},
+                **{record.get("name"): record.get("values")[0]["value"] for record in self._get_insights(ig_media)},
+            }
+
+    def _get_insights(self, item) -> Iterator[Any]:
+        """
+        This is necessary because the functions that call this endpoint return
+        a generator, whose calls need decorated with a backoff.
+        """
+        if item.get("media_type") == "VIDEO":
+            metrics = self.MEDIA_METRICS + ["video_views"]
+        elif item.get("media_type") == "CAROUSEL_ALBUM":
+            metrics = self.CAROUSEL_ALBUM_METRICS
+        else:
+            metrics = self.MEDIA_METRICS
+
+        # An error might occur if the media was posted before the most recent time that
+        # the user's account was converted to a business account from a personal account
+        try:
+            return item.get_insights(params={"metric": metrics})
+        except FacebookRequestError as e:
+            logger.error(f"Insights error: {e.body()}")
+            return []
+
+
+class StoriesInsightsAPI(StoriesAPI):
+    STORY_METRICS = ["exits", "impressions", "reach", "replies", "taps_forward", "taps_back"]
+
+    def list(self, fields: Sequence[str] = None) -> Iterator[dict]:
+        stories = self._get_stories({"limit": self.result_return_limit}, fields=[])
+        for ig_story in stories:
+            yield {
+                **{"id": ig_story.get("id")},
+                **{record.get("name"): record.get("values")[0]["value"] for record in self._get_insights(ig_story)},
```

Comment:
will there always be one value? why do we always index `0` instead of looping over every value in `values`?

Reply:
Yes, for Story/Media Insights there are always be one value. We cannot filter by date, because this is not supported. We receive only the actual Insight at this particular moment from the time of publication of a current post.